include/sys/zpl.h

```diff
@@ -25,6 +25,14 @@
 #ifndef	_SYS_ZPL_H
 #define	_SYS_ZPL_H
 
+/* XXX stuck this here for now just to make it work. posix acl support is
+       controlled by the kernel config option CONFIG_FS_POSIX_ACL, there
+       isn't really anything similar for NFS4 ACL's. The closest thing might
+       be CONFIG_NFS_ACL_SUPPORT, but that doesn't really seem appropriate.
+       Maybe the native zfs acl support shouldn't be conditional? Just always
+       available? */
+#define CONFIG_FS_NFS4_ACL
```

Comment:
Ok; I moved this check to sys/zfs_acl.h, is that a good place or is there somewhere better? CONFIG_KEYS isn't defined; how does CONFIG_FS_POSIX_ACL get defined? I see it checked, but don't see anything in the tree that sets it. I updated the NFS4ACL conditional to ZFS_NFS4_ACL so it doesn't look like a kernel config define, is that good or would something else be better?

Reply:
Both `CONFIG_FS_POSIX_ACL` and `CONFIG_KEYS` _are_ kernel build options.  The ZFS build system inherits the `/usr/src/kernels/<kernel>/include/generated/autoconf.h` include so it's aware of how the kernel was compiled.  This is important since certain symbols may not be available if you kernel was configured without support for them.  For example, when `CONFIG_FS_POSIX_ACL` is undefined none of the Posix ACL functions are built.

Reply:
Right, CONFIG_FS_POSIX_ACL and CONFIG_KEYS are kernel options, but CONFIG_FS_NFS4_ACL (which I made up and initially used) is not. I renamed that to ZFS_NFS4_ACL (which will be conditionally defined based on kernel option checks); my question was whether that's a good name or if you'd rather have something else.

Maybe I'm missing something; I have this code I stuck in zfs_acl.h (for want of a better idea where to put it):

#if defined(CONFIG_KEYS)
#define ZFS_NFS4_ACL
#endif 

I changed it temporarily to:

#if defined(CONFIG_KEYS)
#define ZFS_NFS4_ACL
#else
#error No CONFIG_KEYS?
#endif

and it fails:

In file included from ../../module/zcommon/zfs_prop.c:33:0:
../../include/sys/zfs_acl.h:46:2: error: #error No CONFIG_KEYS?
 #error No CONFIG_KEYS?

So CONFIG_KEYS does not seems to be defined at that point? It is in the kernel config:

$ grep CONFIG_KEYS /usr/src/linux-headers-4.15.0-32-generic/include/generated/autoconf.h
#define CONFIG_KEYS 1
#define CONFIG_KEYS_COMPAT 1

What am I doing wrong?

Reply:
Sure `ZFS_NFS4_ACL` sounds reasonable, and shouldn't conflict with with anything else.

My guess is your getting tripped up by a subtle thing about the layout of the ZFS repository.  Most of the core ZFS code is laid out such that it will be compiled twice.  Once as a kernel module using the kernel build system, and a second time in user space for the `libzpool` shared library.  This second user space build is used by utilities like `zdb` and `ztest` to make it easier to test and verify the code.

You're probably seeing a `libzpool` build failure since the kernel headers are not included.  Many of these source files and headers are built in both contexts so you may need to wrap some sections with `#ifdef _KERNEL` when it doesn't make sense to build it in user space.  You probably didn't run in to this issue before because none of the `zpl_*` files are compiled as part of `libzpool` since they implement the POSIX layer which requires helps from the VFS which doesn't exist in user space.  However, that `zfs_acl.h` header will get included in both contexts.  Hope this helps!

Reply:
Hmm, I could have sworn I compiled without the #error and when I tried to use the resultant kernel module it was giving operation not supported failures as if the feature wasn't compiled in. But now that I try again it seems to be working, so i don't know what happened before.